src/Behat/Behat/Definition/Context/Attribute/DefinitionAttributeReader.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+/*
+ * This file is part of the Behat.
+ * (c) Konstantin Kudryashov <ever.zet@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Behat\Behat\Definition\Context\Attribute;
+
+use Behat\Behat\Context\Attribute\AttributeReader;
+use Behat\Behat\Definition\Attribute\Definition;
+use Behat\Behat\Definition\Attribute\Given;
+use Behat\Behat\Definition\Attribute\Then;
+use Behat\Behat\Definition\Attribute\When;
+use ReflectionMethod;
+
+/**
+ * Reads definition Attributes from the context class.
+ *
+ * @author Konstantin Kudryashov <ever.zet@gmail.com>
+ */
+class DefinitionAttributeReader implements AttributeReader
+{
+    /**
+     * @var string[]
+     */
+    private static $classes = array(
+        Given::class => 'Behat\Behat\Definition\Call\Given',
+        When::class  => 'Behat\Behat\Definition\Call\When',
+        Then::class  => 'Behat\Behat\Definition\Call\Then',
+    );
+
+    /**
+     * @{inheritdoc}
+     */
+    public function readCallees($contextClass, ReflectionMethod $method)
+    {
+        if (PHP_MAJOR_VERSION < 8) {
+            return [];
+        }
+
+        $attributes = $method->getAttributes(Definition::class, \ReflectionAttribute::IS_INSTANCEOF);
+
+        $callees = [];
+        foreach ($attributes as $attribute) {
+            $class = static::$classes[$attribute->getName()];
```

Comment:
accessing private static properties must always use `self::`, not `static::`. Otherwise, child classes are broken, as they try to access `ChildClass::$classes`

Reply:
Nice catch! I've been working with final and abstract classes so long I forgot that was even a thing :slightly_smiling_face:

Reply:
Changed to `self` and made all classes `final` as well.